libs/auth-api-lib/src/lib/services/delegations.service.ts

```diff
@@ -86,19 +90,119 @@ export class DelegationsService {
           fromNationalId: d.fromNationalId,
           fromName: d.fromDisplayName,
-          isFromCompany: d.isFromCompany,
           type: DelegationType.Custom,
           provider: DelegationProvider.Custom,
         },
     )
   }
+
+  async create(
+    nationalId: string,
+    delegation: DelegationDTO,
+  ): Promise<Delegation | null> {
+    this.logger.debug('Creating a new delegation')
+    const id = uuid()
+    await this.delegationModel.create({
+      id: id,
+      fromNationalId: nationalId,
+      ...delegation,
+    })
+    if (delegation.scopes) {
+      this.delegationScopeService.createMany(id, delegation.scopes)
+    }
+    return this.findOne(nationalId, id)
+  }
+
+  async update(
+    nationalId: string,
+    delegation: DelegationDTO,
+    id: string,
+  ): Promise<Delegation | null> {
+    this.logger.debug(`Updating a delegation with id ${id}`)
+    await this.delegationModel.update(
+      { ...delegation },
+      { where: { id: id, fromNationalId: nationalId } },
+    )
+    await this.delegationScopeService.delete(id)
+    if (delegation.scopes) {
+      await this.delegationScopeService.createMany(id, delegation.scopes)
+    }
+    return this.findOne(nationalId, id)
+  }
+
+  async findOne(nationalId: string, id: string): Promise<Delegation | null> {
+    this.logger.debug(`Finding a delegation with id ${id}`)
+    return await this.delegationModel.findOne({
+      where: {
+        id: id,
+        fromNationalId: nationalId,
+      },
+      include: [DelegationScope],
+    })
+  }
+
+  async findAllCustomTo(nationalId: string): Promise<Delegation[] | null> {
+    this.logger.debug(`Finding a delegation for nationalId ${nationalId}`)
+    return await this.delegationModel.findAll({
+      where: {
+        toNationalId: nationalId,
+      },
+      include: [DelegationScope],
+    })
+  }
+
+  async findAllCustomFrom(nationalId: string): Promise<Delegation[] | null> {
+    this.logger.debug(`Finding a delegation for nationalId ${nationalId}`)
+    return await this.delegationModel.findAll({
+      where: {
+        fromNationalId: nationalId,
+      },
+      include: [DelegationScope],
+    })
+  }
+
+  async deleteFrom(nationalId: string, id: string): Promise<number> {
+    this.logger.debug(`Deleting Delegation for Id ${id}`)
+
+    const delegation = await this.delegationModel.findByPk(id)
+    if (!delegation || delegation?.fromNationalId !== nationalId) {
+      this.logger.debug('Delegation does is not assigned to user')
+      return 0
+    }
+
+    await this.delegationScopeService.delete(id)
+
+    const response = this.delegationModel.destroy({
+      where: { id: id, fromNationalId: nationalId },
+    })
+
+    return response
+  }
+
+  async deleteTo(nationalId: string, id: string): Promise<number> {
+    this.logger.debug(`Deleting Delegation for Id ${id}`)
+
+    const delegation = await this.delegationModel.findByPk(id)
+
+    if (!delegation || delegation?.toNationalId !== nationalId) {
```

Comment:
This should most likely be handled in the controller, changing the url to:
```typescript
/users/:nationalId/delegations/:id
```
Then return `404` when not found for that user.

Reply:
Shouldn't at least also be checked in the service so the service is not trusting the controller to take care of this?

Reply:
Is this a showstopper?

Reply:
IMO yes. The consumer of this API (wether it's the GraphQL API, or the client) needs to take some actions based on the HTTP status codes. As it's now, only 200/201 is returned, even though the action wasn't successful.

Reply:
The Api is returning 401 status when the delegation is not owned by the current user. This could easily be 404 or 403 (Forbidden). Whatever is preferred. 
The API therefor returns 401 in this case before trying to delete the record(s).

Reply:
Where is that happening?
e.g.
```typescript
// controller
Promise<number> {
  return await this.delegationsService.deleteFrom(user.nationalId, id)
}
```
```typescript
// service
if (!delegation || delegation?.fromNationalId !== nationalId) {
  this.logger.debug('Delegation does is not assigned to user')
  return 0
}
```

Reply:
I think you are looking at an older version. This is how we are doing this now:

```
if (!delegation || delegation?.fromNationalId !== nationalId) {
      this.logger.debug('Delegation is not assigned to user')
      throw new NotFoundException()
    }
```

Reply:
Geez, sorry. I always clicked on the file link in this conversation where I'm routed to (as you say) the old commit :/.